{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "slideshow": {
     "slide_type": "skip"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\Luca\\Anaconda3\\envs\\bio\\lib\\site-packages\\IPython\\html.py:14: ShimWarning: The `IPython.html` package has been deprecated since IPython 4.0. You should import from `notebook` instead. `IPython.html.widgets` has moved to `ipywidgets`.\n",
      "  \"`IPython.html.widgets` has moved to `ipywidgets`.\", ShimWarning)\n"
     ]
    }
   ],
   "source": [
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "source": [
    "# PyGMQL in action"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "slideshow": {
     "slide_type": "-"
    }
   },
   "source": [
    "## Problem description\n",
    "We are given three replicas of a ChIP-Seq experiment.\n",
    "We want to:\n",
    "1. Extract **high-confidence regions** into one sample\n",
    "2. Identify which of these regions **overlap** with a set of given genes\n",
    "3. For each resulting region count ICGC **mutations**.\n",
    "4. Finally we want to select the regions with **at least one** mutation."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "slideshow": {
     "slide_type": "-"
    }
   },
   "source": [
    "## Required GMQL operations\n",
    "For this pipeline we will need the following GMQL operations:\n",
    "- ```cover```: extracts regions which are confirmed by at least two replicas\n",
    "- ```join```: extracts regions which overalap with genes\n",
    "- ```map```: for each extracted region, counts the overlapping mutations"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "source": [
    "## PyGMQL in a nutshell\n",
    "\n",
    "PyGMQL enables the writing of GMQL queries using the Python programming language. It exposes to the user all the GMQL operators and also a data structure for holding, manipulating and converting the results of a GMQL query."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "source": [
    "The library offers:\n",
    "- A data structure called ```GMQLDataset```, which represent a GMQL variable in the query. Every GMQLDataset is produced by a GMQL operator. When you call the ```materialize``` operation on a GMQLDataset the execution is started and the result is returned.\n",
    "- A data structure called ```GDataframe```, holding the result of a query. A GDataframe is a pure python structure and can be used directly as any other pandas dataframe. A GDataframe holds two pandas dataframes, one for the regions and one for the metadata. We can also, given a GDataframe, go back to a GMQLDataset for using it as a GMQL variable. This can be done calling the ```to_GMQLDataset``` function of the GDataframe.\n",
    "\n",
    "<img src=\"./images/GMQLDataset_to_GDataframe.png\" width=\"500\" height=\"600\">"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Getting the library\n",
    "\n",
    "The library can be downloaded from the PyPi public repository through the ```pip``` packaging system.\n",
    "\n",
    "```\n",
    "    pip install gmql\n",
    "```\n",
    "\n",
    "If you want the most recent version of the software you can directly download it from the GitHub page:\n",
    "\n",
    "```\n",
    "    git pull https://github.com/DEIB-GECO/PyGMQL.git\n",
    "    cd PyGMQL\n",
    "    pip install -e .\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "source": [
    "### Importing the library"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true,
    "slideshow": {
     "slide_type": "-"
    }
   },
   "outputs": [],
   "source": [
    "import gmql as gl\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "source": [
    "### Execution modes\n",
    "The queries are directly embedded inside the language and can be executed in two different modes:\n",
    "- *Local mode*: the computation is executed in the local machine\n",
    "- *Remote mode*: the query is sent to a remote server, executed there and the results are downloaded. This process is shown in the figure below.\n",
    "\n",
    "<img src=\"./images/remote.png\">"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true,
    "slideshow": {
     "slide_type": "-"
    }
   },
   "outputs": [],
   "source": [
    "gl.set_mode(\"local\")\n",
    "gl.set_progress(False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "source": [
    "## The query\n",
    "\n",
    "The GMQL query that we are going to present is the following:\n",
    "\n",
    "```\n",
    "refSeqGenes  = SELECT(annotation_type == 'gene' AND provider == 'RefSeq') HG19_BED_ANNOTATION;\n",
    "\n",
    "myExp  = SELECT() myRawExp;\n",
    "myConfirmExp  = COVER(2, ANY) myExp;\n",
    "\n",
    "myExpOverGenes  = JOIN(DIST < 0; output: RIGHT_DISTINCT) refSeqGenes myConfirmExp;\n",
    "\n",
    "myMut  = SELECT() myRawMut;\n",
    "myMutOverExp = MAP() myExpOverGenes myMut;\n",
    "\n",
    "myFilteredExp = SELECT(region: count_myMutOverExp_myMut > 0) myMutOverExp;\n",
    "\n",
    "MATERIALIZE myFilteredExp INTO ./Results/FilteredExperiment\n",
    "    \n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "source": [
    "## Loading a GMQL dataset"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "source": [
    "### Loading datasets in GDM format\n",
    "\n",
    "Both the gene dataset and the personal one are already in the GDM format, therefore the library only needs the location of the data for importing.\n",
    "\n",
    "*NB: This demo is meant to be executed without the need of a remote server. For this reason the HG19_BED_ANNOTATION dataset is loaded directly from the computer disk. Please be aware that it can also be found in the regular GMQL repository.*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": true,
    "slideshow": {
     "slide_type": "-"
    }
   },
   "outputs": [],
   "source": [
    "bed_annotation = gl.load_from_path(\"./Data/HG19_BED_ANNOTATION/\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": true,
    "slideshow": {
     "slide_type": "-"
    }
   },
   "outputs": [],
   "source": [
    "myExp = gl.load_from_path(\"./Data/myRawExp/\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "source": [
    "### Loading datasets with a generic schema\n",
    "\n",
    "The mutation dataset is in a classical BED format but not in the GDM format, therefore we need to specify its schema through a **custom parser**, which is an instance of a ```RegionParser```. \n",
    "After the parser is instantiated we can use the ```load_from_path``` function to load the dataset, which is currently stored in the local file system"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": true,
    "slideshow": {
     "slide_type": "-"
    }
   },
   "outputs": [],
   "source": [
    "mutations_parser = gl.parsers.RegionParser(parser_name=\"mutations_parser\",\n",
    "                                           chrPos=0,\n",
    "                                           startPos=1,\n",
    "                                           stopPos=2,\n",
    "                                           strandPos=3,\n",
    "                                           delimiter=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": true,
    "slideshow": {
     "slide_type": "-"
    }
   },
   "outputs": [],
   "source": [
    "myMut = gl.load_from_path(\"./Data/myRawMut/\", parser=mutations_parser)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "source": [
    "## Selection on the metadata\n",
    "\n",
    "We have in the variable ```bed_annotation``` all the ENCODE dataset of annotations. We are interested only in the annotations regarding the genes. Therefore we need to *filter* the dataset **on the basis of the metadata**.\n",
    "This is called ```meta-selection``` and in PyGMQL can be performed using the square-bracket notation common to pandas users."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": true,
    "slideshow": {
     "slide_type": "-"
    }
   },
   "outputs": [],
   "source": [
    "refSeqGenes = bed_annotation[(bed_annotation['annotation_type'] == 'gene') & \n",
    "                             (bed_annotation['provider'] == 'RefSeq')]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "source": [
    "## Cover operation\n",
    "We want only reliable data from the Chip-Seq experiment, therefore we define a Chip-Seq region *highly confident* if it is confirmed by at least two replicas. This is a job for the ```cover``` operation.\n",
    "\n",
    "- ```minAcc```: we define the minimum number of overlapping between samples for a region to be conserved. In this case 2.\n",
    "- ```maxAcc```: we define the maximum number. The ```\"ANY\"``` keyword makes the upper bound infinite.\n",
    "\n",
    "<img src=\"./images/cover.PNG\">"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": true,
    "slideshow": {
     "slide_type": "-"
    }
   },
   "outputs": [],
   "source": [
    "myConfirmExp = myExp.normal_cover(minAcc=2, maxAcc=\"ANY\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "source": [
    "## Join\n",
    "Now we want to extract those regions that overlap with genes. We can do it using the ```join``` operation. \n",
    "- We use the genes as reference dataset\n",
    "- The Chip-Seq regions as experiment. \n",
    "- The genometric predicate ```DLE(0)``` and the option ```output=\"RIGHT\"``` tell the engine to look at all the experiment regions that happen to intersect with the reference ones and keep them in the result.\n",
    "\n",
    "<img src=\"./images/join.PNG\" height=\"700\" width=\"700\">"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": true,
    "slideshow": {
     "slide_type": "-"
    }
   },
   "outputs": [],
   "source": [
    "myExpOverGenes = refSeqGenes.join(experiment=myConfirmExp, refName=\"gene\",\n",
    "                                  genometric_predicate=[gl.DLE(0)],\n",
    "                                  output=\"RIGHT_DISTINCT\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "source": [
    "## Map\n",
    "Now in ```insideGene``` we have the set of highly confident Chip-Seq regions that intersect with a gene. In order to see how many mutations happen in those region we can use the ```map``` operation. This operation will add a new attribute to the resulting dataset which will be named ```count_GENE_MUTATION```.\n",
    "\n",
    "<img src=\"./images/map.PNG\">"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": true,
    "slideshow": {
     "slide_type": "-"
    }
   },
   "outputs": [],
   "source": [
    "myMutOverExp = myExpOverGenes.map(myMut, expName=\"MUTATION\", refName=\"GENE\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "source": [
    "## Selection on regions\n",
    "In order to filter out all the regions in ```mutationCount``` that do not have any mutation we can do a selection on region data using the ```reg_select``` operation. This operation takes as input a predicate on region fields."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": true,
    "slideshow": {
     "slide_type": "-"
    }
   },
   "outputs": [],
   "source": [
    "myFilteredExp = myMutOverExp.reg_select(myMutOverExp.count_GENE_MUTATION > 0)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "source": [
    "## Materialization\n",
    "We can now materialize the result. PyGMQL adopts a *lazy loading* approach (like Spark) and no action is performed until materialization. The result can be saved as a GDM dataset and also directly loaded in python as a ```GDataframe```."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "collapsed": true,
    "slideshow": {
     "slide_type": "-"
    }
   },
   "outputs": [],
   "source": [
    "result = myFilteredExp.materialize(\"./Results/FilteredExperiment/\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "source": [
    "## The GDataframe\n",
    "The result of a materialization is a ```GDataframe``` which contains both regions and metadata as two Pandas dataframes ```regs``` and ```meta```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "slideshow": {
     "slide_type": "-"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>strand</th>\n",
       "      <th>accindex</th>\n",
       "      <th>jaccardintersect</th>\n",
       "      <th>jaccardresult</th>\n",
       "      <th>count_gene_mutation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>S_00000.gdm</th>\n",
       "      <td>chr1</td>\n",
       "      <td>8035225</td>\n",
       "      <td>8035786</td>\n",
       "      <td>*</td>\n",
       "      <td>3</td>\n",
       "      <td>0.921182</td>\n",
       "      <td>0.748768</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S_00000.gdm</th>\n",
       "      <td>chr1</td>\n",
       "      <td>11779578</td>\n",
       "      <td>11780328</td>\n",
       "      <td>*</td>\n",
       "      <td>3</td>\n",
       "      <td>0.651607</td>\n",
       "      <td>0.501303</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S_00000.gdm</th>\n",
       "      <td>chr1</td>\n",
       "      <td>11898457</td>\n",
       "      <td>11899199</td>\n",
       "      <td>*</td>\n",
       "      <td>3</td>\n",
       "      <td>0.876033</td>\n",
       "      <td>0.783943</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S_00000.gdm</th>\n",
       "      <td>chr1</td>\n",
       "      <td>17286965</td>\n",
       "      <td>17287470</td>\n",
       "      <td>*</td>\n",
       "      <td>3</td>\n",
       "      <td>0.698479</td>\n",
       "      <td>0.641770</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S_00000.gdm</th>\n",
       "      <td>chr1</td>\n",
       "      <td>22140883</td>\n",
       "      <td>22141897</td>\n",
       "      <td>*</td>\n",
       "      <td>3</td>\n",
       "      <td>0.910233</td>\n",
       "      <td>0.754937</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              chr     start      stop strand  accindex  jaccardintersect  \\\n",
       "S_00000.gdm  chr1   8035225   8035786      *         3          0.921182   \n",
       "S_00000.gdm  chr1  11779578  11780328      *         3          0.651607   \n",
       "S_00000.gdm  chr1  11898457  11899199      *         3          0.876033   \n",
       "S_00000.gdm  chr1  17286965  17287470      *         3          0.698479   \n",
       "S_00000.gdm  chr1  22140883  22141897      *         3          0.910233   \n",
       "\n",
       "             jaccardresult  count_gene_mutation  \n",
       "S_00000.gdm       0.748768                  1.0  \n",
       "S_00000.gdm       0.501303                  1.0  \n",
       "S_00000.gdm       0.783943                  1.0  \n",
       "S_00000.gdm       0.641770                  1.0  \n",
       "S_00000.gdm       0.754937                  2.0  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "result.regs.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "slideshow": {
     "slide_type": "-"
    }
   },
   "source": [
    "The region dataframe represents the regions in the output result in a tabular format. The index of the regions is the identifier of the sample they belong to."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GENE.ID</th>\n",
       "      <th>GENE.antibody</th>\n",
       "      <th>GENE.antibody_antibodyDescription</th>\n",
       "      <th>GENE.antibody_lab</th>\n",
       "      <th>GENE.antibody_label</th>\n",
       "      <th>GENE.antibody_lots</th>\n",
       "      <th>GENE.antibody_orderUrl</th>\n",
       "      <th>GENE.antibody_tag</th>\n",
       "      <th>GENE.antibody_target</th>\n",
       "      <th>GENE.antibody_targetClass</th>\n",
       "      <th>...</th>\n",
       "      <th>GENE.treatment_tag</th>\n",
       "      <th>GENE.treatment_type</th>\n",
       "      <th>GENE.type</th>\n",
       "      <th>GENE.url</th>\n",
       "      <th>GENE.view</th>\n",
       "      <th>GENE.view_description</th>\n",
       "      <th>GENE.view_label</th>\n",
       "      <th>GENE.view_tag</th>\n",
       "      <th>GENE.view_type</th>\n",
       "      <th>MUTATION.type</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>id_sample</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>S_00000.gdm</th>\n",
       "      <td>[1880, 1881, 1882]</td>\n",
       "      <td>[CTCF]</td>\n",
       "      <td>[Rabbit polyclonal. Antibody Target: CTCF]</td>\n",
       "      <td>[Myers, Hardison, Snyder]</td>\n",
       "      <td>[CTCF (07-729)]</td>\n",
       "      <td>[1350637 DAM1472197]</td>\n",
       "      <td>[http://www.millipore.com/catalogue/item/07-729]</td>\n",
       "      <td>[CTCF]</td>\n",
       "      <td>[CTCF]</td>\n",
       "      <td>[TFSS]</td>\n",
       "      <td>...</td>\n",
       "      <td>[NONE]</td>\n",
       "      <td>[control]</td>\n",
       "      <td>[broadPeak]</td>\n",
       "      <td>[http://hgdownload.cse.ucsc.edu/goldenPath/hg1...</td>\n",
       "      <td>[Hotspots]</td>\n",
       "      <td>[ChIP-seq affinity zones identified using the ...</td>\n",
       "      <td>[Hotspots]</td>\n",
       "      <td>[HOTSPOTS]</td>\n",
       "      <td>[view]</td>\n",
       "      <td>[mutations]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows Ã— 92 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                        GENE.ID GENE.antibody  \\\n",
       "id_sample                                       \n",
       "S_00000.gdm  [1880, 1881, 1882]        [CTCF]   \n",
       "\n",
       "                      GENE.antibody_antibodyDescription  \\\n",
       "id_sample                                                 \n",
       "S_00000.gdm  [Rabbit polyclonal. Antibody Target: CTCF]   \n",
       "\n",
       "                     GENE.antibody_lab GENE.antibody_label  \\\n",
       "id_sample                                                    \n",
       "S_00000.gdm  [Myers, Hardison, Snyder]     [CTCF (07-729)]   \n",
       "\n",
       "               GENE.antibody_lots  \\\n",
       "id_sample                           \n",
       "S_00000.gdm  [1350637 DAM1472197]   \n",
       "\n",
       "                                       GENE.antibody_orderUrl  \\\n",
       "id_sample                                                       \n",
       "S_00000.gdm  [http://www.millipore.com/catalogue/item/07-729]   \n",
       "\n",
       "            GENE.antibody_tag GENE.antibody_target GENE.antibody_targetClass  \\\n",
       "id_sample                                                                      \n",
       "S_00000.gdm            [CTCF]               [CTCF]                    [TFSS]   \n",
       "\n",
       "                 ...      GENE.treatment_tag GENE.treatment_type    GENE.type  \\\n",
       "id_sample        ...                                                            \n",
       "S_00000.gdm      ...                  [NONE]           [control]  [broadPeak]   \n",
       "\n",
       "                                                      GENE.url   GENE.view  \\\n",
       "id_sample                                                                    \n",
       "S_00000.gdm  [http://hgdownload.cse.ucsc.edu/goldenPath/hg1...  [Hotspots]   \n",
       "\n",
       "                                         GENE.view_description  \\\n",
       "id_sample                                                        \n",
       "S_00000.gdm  [ChIP-seq affinity zones identified using the ...   \n",
       "\n",
       "            GENE.view_label GENE.view_tag GENE.view_type MUTATION.type  \n",
       "id_sample                                                               \n",
       "S_00000.gdm      [Hotspots]    [HOTSPOTS]         [view]   [mutations]  \n",
       "\n",
       "[1 rows x 92 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "result.meta"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "slideshow": {
     "slide_type": "-"
    }
   },
   "source": [
    "The metadata dataframe has as columns all the metadata attributes and every row represents a sample in the output dataset. \n",
    "\n",
    "There can be multiple values in the same cell."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "source": [
    "## Additional operations\n",
    "Having the result in a Pandas dataframe enables us to use all the Pandas functions and perform some statistics.\n",
    "\n",
    "For example we can do an histogram of the mutation counts over the Chip-Seq regions showing the distribution of the number of mutations over those regions."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "slideshow": {
     "slide_type": "-"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[<matplotlib.axes._subplots.AxesSubplot object at 0x000001FF7CDAF668>]], dtype=object)"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "<matplotlib.figure.Figure at 0x1ff7ce0cc88>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAXUAAAEFCAYAAAAc33cJAAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAEvFJREFUeJzt3X+QXWV9x/H3kiVJo0u6yiq0Q0WLftV2QCdWKDQkIr8F\nQlFLpzr+QBGnQVDT8jNoR6FIi6hoKTZIo46OVH4qnYgtCI1UpVi0RuHLQGV0itotbmAxgoRs/zgn\nctnJ7t49e3dv9rnv1z+cc+65z/N9LvA5zz7n/ugbGxtDklSGXbpdgCSpcwx1SSqIoS5JBTHUJakg\nhrokFcRQl6SCGOraaUXEVyNi927XMVciYl1ELJvOeRFxRUQcOvvVab4w1LUzO6zbBcyxw4C+6ZyX\nmW/PzH+d1ao0r/T54SM1EREnAWuAJ4H/A94MHAWcVh/7GXBqZt4bEeuBTZl5cf3cX+9HxAPAeuDV\nwO8AV2XmGRHxj8BbgE3A0Zn540lqOQt4GzAK/BtwfGbuHRELgYuAFcAC4C7gtMx8ZKJ+6/aOBdYC\nC4EtwF9k5jemeD0eAD4PvAZ4NvB+4CBgGfAEcFxmPlif97rMvLPlea8D/hj4S+CHwJuoQvtvgEXA\nnsC/ZObbIuKCceddBHwiM6+OiOPrfhcAjwDvzcw7IuKvgL3rdp4HDAMnZuaDk41J85MzdU1bROxH\nFSZHZua+wJeAm4EzgFdl5n5UAXd9RLQz83xmZi4HDgTeFRHPz8y31o+9aopAP4Iq/P+AKkAHWh4+\nC9gKLKtrehD40GT9RsQLgb+mupC8HHgHcG1EPKONcSyu+1kD/APwsXr/x3WNE8rMc+v63pCZ3wJO\nB96XmfsDLwWOi4hlOzhv++vwYuBy4LX1v5P3ATdExG71KcuB12fmi4ER4JQ2xqN5yFBXE68Gbtoe\ntpn5UeB6qtnucH1sPfDbVDPEqdxQP+d/gP8FnjWNWo4GvpiZmzNzDPi7lseOAVYBd0XEd4DjqQJy\nsn4Po5rR3lw/53PANmCfNmq5pv7n/cBPM/O7LfvTGRNUf/n8ZkScA1wGLAGeOcn5hwA3Z+Z/A2Tm\nLVRj2r5Gf2tmPlJv39WgHs0T/d0uQPPSVuDX63YR8RvAC4Acd14fsGt9buuMfeG4837Zsj3+3HZq\naT3/yZbtBcDpmbmhrvOZwOIp+l1AFY4nbn8gIvaimh1P5fGW7ScmOGeq12K7jcB3ga8A/wTsz+Sv\ny44maLtQvf4ws9dY84gzdTXxNeDQiNiz3j+Fai35xIgYAoiItwIPAfdRreG+oj6+O9VSQDue5KlQ\nmsg/A6+NiKX1/tt46oJzE3BqRCyMiF2AdcCFU7R3C3B4vZxBRBwN/BdPvxjMROtrcQDVXwXbbQV2\njYjB+pwzM/Naqr949qG64Pz6vAnqfkHd9iHAXsC3UE8x1DVtmfk9qpt1X4mI7wJHAr8LfAS4JSK+\nT7V8cExmbgM+DuwZEUm1nHFrm11dC3w9In5/klpuoQrrb0TEncBSqpubAB8EHqBabvgB1ex0zRRj\n+z7VOvoX6rF9kOom5y/arHkqZwKn10s7JwPfbnnseuAqqvsDFwL/WY/pbOB2nloCuh64KiIOb6n7\nB8CfU63/b6K6d3BsZj7cobo1T/juF81rEfEK4MDMvLTefy+wf+vyidRLXFPXTi8irgJigoffDCyP\niHdQLbv8iGqm3eka3kD118mOfC4z/7bTfUpNOFOXpIK4pi5JBTHUJakgXV1THx4enZdrP4ODSxgZ\n2TL1iQXptTH32njBMc8nQ0MDE37OwJl6A/39C6Y+qTC9NuZeGy845lIY6pJUEENdkgpiqEtSQQx1\nSSqIoS5JBTHUJakgbYV6ROwfEbeOO/ZnEfGNlv2TI+LOiPhmRBzT4TolSW2YMtQj4gzgClq+Tzoi\nXk71vdV99f4eVL9NeRBwBHBhRCyajYIlSRNrZ6Z+P3DC9p2IeDbVbzi+u+WcVwK3Z+bj9fc33wfs\n28lCJUlTm/JrAjLzmojYGyAiFgCfAt7L038eazeg9cv4R6l+rGBSg4NLGn+i69g1N+zw+Jc/vKpR\ne9M1NDQw9UmF6bUx99p4wTGXYLrf/bIMeCHw91TLMS+NiI9S/ZRW6yszAGyeqrHZ+M6F4eHRjrc5\n3tDQwJz0szPptTH32njBMc8nk12IphXqmXkH8HsA9ez9C5n57npN/YKIWAwsAl4CbGpasCSpmY68\npTEzfwpcSvUL6LcA52bmY51oW5LUvrZm6pn5AHDAZMcycx3VDwBLkrrEDx9JUkEMdUkqiKEuSQUx\n1CWpIIa6JBXEUJekghjqklQQQ12SCmKoS1JBDHVJKoihLkkFMdQlqSCGuiQVxFCXpIIY6pJUEENd\nkgpiqEtSQQx1SSqIoS5JBTHUJakghrokFcRQl6SC9LdzUkTsD1yUmSsj4mXAx4EngceBN2XmzyLi\nZOAUYCtwfmbeOFtFS5J2bMqZekScAVwBLK4PfQx4V2auBK4FzoyIPYDTgIOAI4ALI2LRrFQsSZpQ\nOzP1+4ETgM/W+3+amT9pef5jwCuB2zPzceDxiLgP2Bf4j8kaHhxcQn//gkaFT2RoaKCj7XW7n51J\nr42518YLjrkEU4Z6Zl4TEXu37P8EICIOBE4FDqaanT/c8rRRYOlUbY+MbJlmuVMbHh7teJvjDQ0N\nzEk/O5NeG3OvjRcc83wy2YWo0Y3SiDgRuBx4TWYOA48Arb0MAJubtC1Jaq6tG6WtIuKNVDdEV2bm\nz+vDdwAXRMRiYBHwEmBTx6qUJLVlWqEeEQuAS4EfAddGBMBtmfn+iLgU2Eg1+z83Mx/rdLGSpMm1\nFeqZ+QBwQL37rAnOWQes60xZkqQm/PCRJBXEUJekghjqklQQQ12SCmKoS1JBDHVJKoihLkkFMdQl\nqSCGuiQVxFCXpIIY6pJUEENdkgpiqEtSQQx1SSqIoS5JBTHUJakghrokFcRQl6SCGOqSVBBDXZIK\nYqhLUkH62zkpIvYHLsrMlRGxD7AeGAM2Aaszc1tEnAycAmwFzs/MG2epZknSBKacqUfEGcAVwOL6\n0CXA2sxcDvQBqyJiD+A04CDgCODCiFg0OyVLkibSzvLL/cAJLfvLgNvq7Q3AocArgdsz8/HMfBi4\nD9i3k4VKkqY25fJLZl4TEXu3HOrLzLF6exRYCuwGPNxyzvbjkxocXEJ//4L2q23D0NBAR9vrdj87\nk14bc6+NFxxzCdpaUx9nW8v2ALAZeKTeHn98UiMjWxp0P7nh4dGOtzne0NDAnPSzM+m1MffaeMEx\nzyeTXYiavPvlrohYWW8fBWwE7gCWR8TiiFgKvITqJqokaQ41mamvAdZFxELgbuDqzHwyIi6lCvhd\ngHMz87EO1ilJakNboZ6ZDwAH1Nv3Ait2cM46YF0ni5MkTY8fPpKkghjqklQQQ12SCmKoS1JBDHVJ\nKoihLkkFMdQlqSCGuiQVxFCXpIIY6pJUEENdkgpiqEtSQQx1SSqIoS5JBTHUJakghrokFcRQl6SC\nGOqSVBBDXZIKYqhLUkEMdUkqiKEuSQXpb/KkiNgV+DSwN/AkcDKwFVgPjAGbgNWZua0jVUqS2tJ0\npn400J+ZBwIfAC4ALgHWZuZyoA9Y1ZkSJUntajRTB+4F+iNiF2A34AngAOC2+vENwOHAdZM1Mji4\nhP7+BQ1L2LGhoYGOttftfnYmvTbmXhsvOOYSNA31R6mWXu4BdgeOAQ7OzLH68VFg6VSNjIxsadj9\nxIaHRzve5nhDQwNz0s/OpNfG3GvjBcc8n0x2IWq6/PIe4KbMfBGwH9X6+sKWxweAzQ3bliQ11DTU\nR4CH6+2fA7sCd0XEyvrYUcDGmZUmSZqupssvHwGujIiNVDP0c4A7gXURsRC4G7i6MyVKktrVKNQz\n81HgT3bw0IqZlSNJmgk/fCRJBTHUJakghrokFcRQl6SCGOqSVBBDXZIKYqhLUkEMdUkqiKEuSQUx\n1CWpIIa6JBXEUJekghjqklQQQ12SCmKoS1JBDHVJKoihLkkFMdQlqSCGuiQVxFCXpIIY6pJUkP6m\nT4yIs4HjgIXAZcBtwHpgDNgErM7MbR2oUZLUpkYz9YhYCRwIHASsAPYCLgHWZuZyoA9Y1aEaJUlt\narr8cgTwPeA64MvAjcAyqtk6wAbg0BlXJ0malqbLL7sDzwOOAZ4PfAnYJTPH6sdHgaVTNTI4uIT+\n/gUNS9ixoaGBjrbX7X52Jr025l4bLzjmEjQN9YeAezLzV0BGxGNUSzDbDQCbp2pkZGRLw+4nNjw8\n2vE2xxsaGpiTfnYmvTbmXhsvOOb5ZLILUdPll68DR0ZEX0T8FvAM4OZ6rR3gKGBjw7YlSQ01mqln\n5o0RcTBwB9WFYTXwQ2BdRCwE7gau7liVkqS2NH5LY2aesYPDK2ZQiyRphvzwkSQVxFCXpIIY6pJU\nEENdkgpiqEtSQQx1SSqIoS5JBTHUJakghrokFcRQl6SCGOqSVBBDXZIKYqhLUkEMdUkqiKEuSQUx\n1CWpIIa6JBXEUJekghjqklQQQ12SCmKoS1JBDHVJKkj/TJ4cEc8Bvg0cBmwF1gNjwCZgdWZum2mB\nkqT2NZ6pR8SuwCeBX9aHLgHWZuZyoA9YNfPyJEnTMZPll4uBy4EH6/1lwG319gbg0Bm0LUlqoNHy\nS0S8BRjOzJsi4uz6cF9mjtXbo8DSqdoZHFxCf/+CJiVMaGhooKPtdbufnUmvjbnXxguOuQRN19RP\nAsYi4lDgZcBngOe0PD4AbJ6qkZGRLQ27n9jw8GjH2xxvaGhgTvrZmfTamHttvOCY55PJLkSNll8y\n8+DMXJGZK4HvAG8CNkTEyvqUo4CNTdqWJDU3o3e/jLMGWBcRC4G7gas72LYkqQ0zDvV6tr7dipm2\nJ0lqzg8fSVJBDHVJKoihLkkFMdQlqSCGuiQVxFCXpIIY6pJUEENdkgpiqEtSQQx1SSqIoS5JBTHU\nJakghrokFcRQl6SCGOqSVBBDXZIKYqhLUkEMdUkqiKEuSQUx1CWpIDP+4eledOyaG3Z4/MqzDpnj\nSiTp6ZypS1JBGs3UI2JX4Epgb2ARcD7wA2A9MAZsAlZn5raOVClJakvTmfobgYcyczlwJPAJ4BJg\nbX2sD1jVmRIlSe1quqb+ReDqersP2AosA26rj20ADgeum6yRwcEl9PcvaFjCjg0NDXS0vfnS91wo\nfXzj9dp4wTGXoFGoZ+ajABExQBXua4GLM3OsPmUUWDpVOyMjW5p0P6nh4dGOtzkf+p5tQ0MDRY9v\nvF4bLzjm+WSyC1HjG6URsRfwNeCzmfl5oHX9fADY3LRtSVIzjUI9Ip4LfBU4MzOvrA/fFREr6+2j\ngI0zL0+SNB1N19TPAQaB8yLivPrY6cClEbEQuJun1twlSXOk6Zr66VQhPt6KmZUjSZoJP3wkSQUx\n1CWpIIa6JBXEUJekghjqklQQQ12SCmKoS1JBDHVJKoihLkkFMdQlqSCGuiQVxFCXpIIY6pJUkKZf\nvasuOOlDt+zw+JVnHTLHlUjaWTlTl6SCGOqSVBBDXZIKYqhLUkG8Uaq2HLvmhh0e9yattHNxpi5J\nBTHUJakgHV1+iYhdgMuA/YDHgbdn5n2d7EO9xffmS9PT6TX144HFmfmHEXEA8GFgVYf7kGbdRPcQ\nwAuKpmeuJyadXn75I+ArAJn5TeAVHW5fkjSJvrGxsY41FhFXANdk5oZ6/0fACzJza8c6kSRNqNMz\n9UeAgdb2DXRJmjudDvXbgaMB6jX173W4fUnSJDp9o/Q64LCI+HegD3hrh9uXJE2io2vqkqTu8sNH\nklQQQ12SCmKoS1JB/JbGNkXErsCVwN7AIuD8zPxSV4uaIxHxHODbwGGZeU+365ltEXE2cBywELgs\nMz/V5ZJmVf3f9qep/tt+Eji55H/PEbE/cFFmroyIfYD1wBiwCVidmdu6Wd9MOVNv3xuBhzJzOXAk\n8Iku1zMn6v/hPwn8stu1zIWIWAkcCBwErAD26mpBc+NooD8zDwQ+AFzQ5XpmTUScAVwBLK4PXQKs\nrf+/7qOArzUx1Nv3ReC8ersP6JUPVV0MXA482O1C5sgRVJ+vuA74MnBjd8uZE/cC/fUX8u0GPNHl\nembT/cAJLfvLgNvq7Q3AoXNeUYcZ6m3KzEczczQiBoCrgbXdrmm2RcRbgOHMvKnbtcyh3am+s+j1\nwDuBz0VEX3dLmnWPUi293AOsAy7tajWzKDOv4ekXrb7M3P6+7lFg6dxX1VmG+jRExF7A14DPZubn\nu13PHDiJ6sNktwIvAz4TEXt0t6RZ9xBwU2b+KjMTeAwY6nJNs+09VGN+EdXXZn86IhZP8ZxStK6f\nDwCbu1VIp3ijtE0R8Vzgq8CpmXlzt+uZC5l58PbtOtjfmZk/7V5Fc+LrwOkRcQmwJ/AMqqAv2QhP\nzV5/DuwKLOheOXPqrohYmZm3AkdRTdrmNUO9fecAg8B5EbF9bf2ozOyJG4i9IjNvjIiDgTuo/pJd\nnZlPdrms2fYR4MqI2Ej1jp9zMvMXXa5prqwB1kXEQuBuqqXVec2vCZCkgrimLkkFMdQlqSCGuiQV\nxFCXpIIY6pJUEENdkgpiqEtSQf4fYVPjqTPzJN0AAAAASUVORK5CYII=\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x1ff7ce0c908>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "plt.figure(figsize=(25, 20))\n",
    "result.regs[result.regs.count_gene_mutation <=50].hist(\"count_gene_mutation\", bins=50)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "source": [
    "Of course we can also simply display the mean value"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "slideshow": {
     "slide_type": "-"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1.404494382022472"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "result.regs['count_gene_mutation'].mean()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "source": [
    "### High density regions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "slideshow": {
     "slide_type": "-"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>strand</th>\n",
       "      <th>accindex</th>\n",
       "      <th>jaccardintersect</th>\n",
       "      <th>jaccardresult</th>\n",
       "      <th>count_gene_mutation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>S_00000.gdm</th>\n",
       "      <td>chr19</td>\n",
       "      <td>19296510</td>\n",
       "      <td>19297234</td>\n",
       "      <td>*</td>\n",
       "      <td>3</td>\n",
       "      <td>0.860880</td>\n",
       "      <td>0.764566</td>\n",
       "      <td>11.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S_00000.gdm</th>\n",
       "      <td>chr19</td>\n",
       "      <td>49495965</td>\n",
       "      <td>49497004</td>\n",
       "      <td>*</td>\n",
       "      <td>3</td>\n",
       "      <td>0.913808</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>6.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               chr     start      stop strand  accindex  jaccardintersect  \\\n",
       "S_00000.gdm  chr19  19296510  19297234      *         3          0.860880   \n",
       "S_00000.gdm  chr19  49495965  49497004      *         3          0.913808   \n",
       "\n",
       "             jaccardresult  count_gene_mutation  \n",
       "S_00000.gdm       0.764566                 11.0  \n",
       "S_00000.gdm       0.000000                  6.0  "
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "result.regs[result.regs.count_gene_mutation > 5].sort_values(\"count_gene_mutation\", ascending=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "slideshow": {
     "slide_type": "skip"
    }
   },
   "source": [
    "## Alternative experiment\n",
    "Mapping the mutations over the genes instead that over the experiment regions\n",
    "\n",
    "\n",
    "The GMQL query that we are going to present is the following:\n",
    "\n",
    "```\n",
    "refSeqGenes  = SELECT(annotation_type == 'gene' AND provider == 'RefSeq') HG19_BED_ANNOTATION;\n",
    "\n",
    "myExp  = SELECT() myRawExp;\n",
    "myConfirmExp  = COVER(2, ANY) myExp;\n",
    "\n",
    "myExpOverGenes  = JOIN(DIST < 0; output: RIGHT_DISTINCT) refSeqGenes myConfirmExp;\n",
    "\n",
    "myMut  = SELECT() myRawMut;\n",
    "myMutOverExp = MAP() myExpOverGenes myMut;\n",
    "\n",
    "myFilteredExp = SELECT(region: count_myMutOverExp_myMut > 0) myMutOverExp;\n",
    "\n",
    "MATERIALIZE myFilteredExp INTO ./Results/FilteredExperiment\n",
    "    \n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "genesOverExp = refSeqGenes.join(experiment=myConfirmExp, refName=\"gene\",\n",
    "                                  genometric_predicate=[gl.DLE(0)],\n",
    "                                  output=\"LEFT_DISTINCT\")\n",
    "myMutOverGenes = genesOverExp.map(myMut, expName=\"MUTATION\", refName=\"GENE\")\n",
    "myFilteredGenes = myMutOverGenes.reg_select(myMutOverGenes.count_GENE_MUTATION > 0)\n",
    "result_genes = myFilteredGenes.materialize(\"./Results/FilteredGenes/\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>strand</th>\n",
       "      <th>name</th>\n",
       "      <th>score</th>\n",
       "      <th>count_gene_mutation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>S_00000.gdm</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1716723</td>\n",
       "      <td>1822556</td>\n",
       "      <td>-</td>\n",
       "      <td>NM_002074</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S_00000.gdm</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1716723</td>\n",
       "      <td>1822556</td>\n",
       "      <td>-</td>\n",
       "      <td>NM_001282539</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S_00000.gdm</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1716723</td>\n",
       "      <td>1822556</td>\n",
       "      <td>-</td>\n",
       "      <td>NM_001282538</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S_00000.gdm</th>\n",
       "      <td>chr1</td>\n",
       "      <td>2115898</td>\n",
       "      <td>2139172</td>\n",
       "      <td>-</td>\n",
       "      <td>NM_001146310</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S_00000.gdm</th>\n",
       "      <td>chr1</td>\n",
       "      <td>2120987</td>\n",
       "      <td>2144159</td>\n",
       "      <td>-</td>\n",
       "      <td>NM_001282670</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              chr    start     stop strand          name  score  \\\n",
       "S_00000.gdm  chr1  1716723  1822556      -     NM_002074    0.0   \n",
       "S_00000.gdm  chr1  1716723  1822556      -  NM_001282539    0.0   \n",
       "S_00000.gdm  chr1  1716723  1822556      -  NM_001282538    0.0   \n",
       "S_00000.gdm  chr1  2115898  2139172      -  NM_001146310    0.0   \n",
       "S_00000.gdm  chr1  2120987  2144159      -  NM_001282670    0.0   \n",
       "\n",
       "             count_gene_mutation  \n",
       "S_00000.gdm                  2.0  \n",
       "S_00000.gdm                  2.0  \n",
       "S_00000.gdm                  2.0  \n",
       "S_00000.gdm                  2.0  \n",
       "S_00000.gdm                  2.0  "
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "result_genes.regs.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[<matplotlib.axes._subplots.AxesSubplot object at 0x000001FF7D1614A8>]], dtype=object)"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "<matplotlib.figure.Figure at 0x1ff7d161b00>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAXoAAAEFCAYAAADt1CyEAAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAE3tJREFUeJzt3X+QXWV9x/H3kiUJqUu6rRvRDpVa7FdqB+zEFmuIpMiv\npCL4gzrjr0gUkGLBkZafwZYRilCkbbSiDcaA6JQaKHboBGiNIqRUBkVLhPkysTI6ZWxXJpDVSDTJ\n9o9zFq6Z3b13f91kn/N+zTCc85zn3vt8N/A5zz7nnpOe4eFhJEnlOmBfD0CSNLMMekkqnEEvSYUz\n6CWpcAa9JBXOoJekwhn0mnUi4p6IeOG+Hke3RMTaiFg8kX4RcWNEHD/zo9NsYNBrNjphXw+gy04A\neibSLzPfl5n/PqOj0qzR4w1Tmk4RsQq4ANgN/AhYCSwHzqvb/hf4QGY+HhHrgS2ZeV392uf2I+IJ\nYD3weuDXgVsz88KI+CzwHmALsCIzfzDOWC4G3gsMAV8DTsvMwyJiLnANcCwwB3gYOC8zt4/1ufX7\nnQKsBuYCO4A/y8wH2vw8ngC+APwR8KvAXwBLgMXAz4E3ZuaTdb+3ZuZDLa97K/Am4M+B7wHvpgry\na4F5wIuBf8vM90bEVXv1uwb4RGZuiIjT6s+dA2wHPpSZD0bEXwKH1e/zUmAQeFtmPjleTZp9nNFr\n2kTEUVQBc3JmHgn8C/Bl4ELgDzPzKKrQuyMiOpmhviAzlwKvBf40In4jM8+oj/1hm5A/ieqE8HtU\nodrXcvhiYBewuB7Tk8BHx/vciHg58FdUJ5ffBc4Cbo+IX+qgjvn151wA/APwd/X+D+oxjikzL6vH\n947M/DpwPvDhzDwa+G3gjRGxeJR+Iz+HVwCfAt5S/5l8GPhSRBxcd1kKnJ6ZrwC2AWd3UI9mGYNe\n0+n1wN0jAZyZfwvcQTUrHqzb1gO/RjWTbOdL9Wv+B/g/4FcmMJYVwBcz8+nMHAb+vuXYG4BTgYcj\n4lvAaVShOd7nnkA18/1y/ZrPA3uAwzsYy231v78L/DAzv92yP5GaoPoN6Zcj4lLgk8AC4AXj9D8O\n+HJm/jdAZm6iqmlkzf+rmbm93n54EuPRLNC7rwegouwCnlsLjIiDgJcBuVe/HuDAum/rzH7uXv1+\n2rK9d99OxtLaf3fL9hzg/MzcWI/zBcD8Np87hyow3zZyICIOpZpFt7OzZfvnY/Rp97MYcR/wbeAu\n4J+Aoxn/5zLaZO4Aqp8/TO1nrFnCGb2m01eA4yPixfX+2VRr02+LiAGAiDgDeArYSrUm/Oq6/YVU\nywid2M3zQTWWfwXeEhEL6/338vxJ6G7gAxExNyIOANYCV7d5v03AifVSCBGxAvgvfvEEMRWtP4vX\nUP32MGIXcGBE9Nd9LsrM26l+Mzqc6iT0XL8xxv2y+r2PAw4Fvo4aw6DXtMnMR6guCN4VEd8GTgZ+\nE/gbYFNEfIdq6eENmbkH+Djw4ohIqqWQr3b4UbcD90fE74wzlk1UAf5ARDwELKS6gArwEeAJqqWK\nR6lmsRe0qe07VOvy/1jX9hGqC6k/6XDM7VwEnF8vC50JfKPl2B3ArVTXG64GvlnXdAmwmeeXj+4A\nbo2IE1vG/SjwJ1TXE7ZQXYs4JTOfmaZxaxbwWzcqUkS8GnhtZq6p9z8EHN269CI1hWv0mrUi4lYg\nxji8ElgaEWdRLdl8n2pGPt1jeAfVbzGj+Xxm/vV0f6Y0Uc7oJalwrtFLUuEMekkq3H65Rj84ONR2\nPam/fwHbtu1o161I1t7M2qHZ9Te5duis/oGBvlHvg5i1M/re3jntOxXK2puryfU3uXaYWv0dzegj\n4ptUD0OC6qFJV1E9+GmY6uFS52bmnog4k+ommV3AlZl5Z3135C3AIqqHS60cuR1ekjTz2s7oI2I+\n0JOZy+p/zgCuB1bXD37qAU6NiEOonlC4BDgJuDoi5gHnAI/UfW+mevqfJKlLOpnRHwUsiIh76v6X\nUj0Q6d76+EbgRKrb0jdn5k5gZ0RsBY4EjqF6rOpI38unb/iSpHY6CfodwHXAjcDLqcK6p34iIFTL\nMQuBg4HW26pHax9pG1d//4KO1qMGBvra9imVtTdXk+tvcu0w+fo7CfrHga11sD8eEU/x/CNOoXrO\n99NUa/h9bdpH2sbVyZX1gYE+BgeHOhh+eay9mbVDs+tvcu3QWf1jnQg6+dbNKuBjABHxEqoZ+j0R\nsaw+vpzq0akPUt1yPr9+YuARVBdqN1M9G7y1rySpSzqZ0X8GWB8R91N9y2YV1V8Rt7b+K9keAzZk\n5u6IWEMV5AcAl2XmsxFxA3BT/fqfAW+fiUIkSaPbL59108kNU03+Nc7am1k7NLv+JtcOHS/dlHXD\nlCSpM/vlIxCmYtVHN43avu7i47o8EknaPzijl6TCGfSSVDiDXpIKZ9BLUuEMekkqnEEvSYUz6CWp\ncAa9JBXOoJekwhn0klQ4g16SCmfQS1LhDHpJKpxBL0mFM+glqXAGvSQVzqCXpMIZ9JJUOINekgpn\n0EtS4Qx6SSqcQS9JhTPoJalwBr0kFc6gl6TCGfSSVDiDXpIKZ9BLUuEMekkqnEEvSYUz6CWpcAa9\nJBXOoJekwvV20ikiFgHfAE4AdgHrgWFgC3BuZu6JiDOBs+vjV2bmnRFxEHALsAgYAlZm5uC0VyFJ\nGlPbGX1EHAh8Gvhp3XQ9sDozlwI9wKkRcQhwHrAEOAm4OiLmAecAj9R9bwZWT38JkqTxdDKjvw74\nFHBJvb8YuLfe3gicCOwGNmfmTmBnRGwFjgSOAa5t6Xt5J4Pq719Ab++ctv0GBvo6ebsJ950NSqtn\nIppcOzS7/ibXDpOvf9ygj4j3AIOZeXdEjAR9T2YO19tDwELgYOCZlpeO1j7S1ta2bTva9hkY6GNw\ncKiTtwOYUN/93URrL0mTa4dm19/k2qGz+sc6EbSb0a8ChiPieOBVVMsvi1qO9wFPA9vr7fHaR9ok\nSV007hp9Zr4uM4/NzGXAt4B3AxsjYlndZTlwH/AgsDQi5kfEQuAIqgu1m4EVe/WVJHXRZL5eeQFw\nRUQ8AMwFNmTmD4E1VEG+CbgsM58FbgBeGRH3A2cBV0zPsCVJnero65UA9ax+xLGjHF8LrN2rbQdw\n+mQHJ0maOm+YkqTCGfSSVDiDXpIKZ9BLUuEMekkqnEEvSYUz6CWpcAa9JBXOoJekwhn0klQ4g16S\nCmfQS1LhDHpJKpxBL0mFM+glqXAGvSQVzqCXpMIZ9JJUOINekgpn0EtS4Qx6SSqcQS9JhTPoJalw\nBr0kFc6gl6TCGfSSVDiDXpIKZ9BLUuEMekkqnEEvSYUz6CWpcAa9JBXOoJekwhn0klS43nYdImIO\nsBYIYBh4P/AssL7e3wKcm5l7IuJM4GxgF3BlZt4ZEQcBtwCLgCFgZWYOzkAtkqRRdDKjPwUgM5cA\nq4GrgOuB1Zm5FOgBTo2IQ4DzgCXAScDVETEPOAd4pO57c/0ekqQuaRv0mXkHcFa9+1LgaWAxcG/d\nthE4Hvh9YHNm7szMZ4CtwJHAMcBde/WVJHVJ26UbgMzcFRE3AW8C3gqckJnD9eEhYCFwMPBMy8tG\nax9pG1d//wJ6e+e0HdfAQF8nw59w39mgtHomosm1Q7Prb3LtMPn6Owp6gMxcGREXAV8HDmo51Ec1\ny99eb4/XPtI2rm3bdrQdz8BAH4ODQx2NHZhQ3/3dRGsvSZNrh2bX3+TaobP6xzoRtF26iYh3RcQl\n9e4OYA/wUEQsq9uWA/cBDwJLI2J+RCwEjqC6ULsZWLFXX0lSl3Qyo78d+GxEfA04EPgg8BiwNiLm\n1tsbMnN3RKyhCvIDgMsy89mIuAG4KSLuB34GvH0mCpEkja5t0GfmT4A/HuXQsaP0XUv1VczWth3A\n6ZMdoCRparxhSpIKZ9BLUuEMekkqnEEvSYUz6CWpcAa9JBXOoJekwhn0klQ4g16SCmfQS1LhDHpJ\nKpxBL0mFM+glqXAGvSQVzqCXpMIZ9JJUOINekgpn0EtS4Qx6SSqcQS9JhTPoJalwBr0kFc6gl6TC\nGfSSVDiDXpIKZ9BLUuEMekkqnEEvSYUz6CWpcAa9JBXOoJekwhn0klQ4g16SCmfQS1LhDHpJKlzv\neAcj4kBgHXAYMA+4EngUWA8MA1uAczNzT0ScCZwN7AKuzMw7I+Ig4BZgETAErMzMwZkpRZI0mnYz\n+ncCT2XmUuBk4BPA9cDquq0HODUiDgHOA5YAJwFXR8Q84BzgkbrvzcDqmSlDkjSWcWf0wBeBDfV2\nD9VsfTFwb922ETgR2A1szsydwM6I2AocCRwDXNvS9/JOBtXfv4De3jlt+w0M9HXydhPuOxuUVs9E\nNLl2aHb9Ta4dJl//uEGfmT8GiIg+qsBfDVyXmcN1lyFgIXAw8EzLS0drH2lra9u2HW37DAz0MTg4\n1MnbAXDKBV8atX3dxcd1/B77i4nWXpIm1w7Nrr/JtUNn9Y91Imh7MTYiDgW+AnwuM78A7Gk53Ac8\nDWyvt8drH2mTJHXRuEEfES8C7gEuysx1dfPDEbGs3l4O3Ac8CCyNiPkRsRA4gupC7WZgxV59JUld\n1G6N/lKgH7g8IkbW188H1kTEXOAxYENm7o6INVRBfgBwWWY+GxE3ADdFxP3Az4C3z0gVkqQxtVuj\nP58q2Pd27Ch91wJr92rbAZw+lQFKkqbGG6YkqXAGvSQVzqCXpMIZ9JJUOINekgpn0EtS4Qx6SSqc\nQS9JhTPoJalwBr0kFc6gl6TCGfSSVDiDXpIKZ9BLUuEMekkqnEEvSYUz6CWpcAa9JBXOoJekwhn0\nklQ4g16SCmfQS1LhDHpJKpxBL0mFM+glqXAGvSQVzqCXpMIZ9JJUOINekgpn0EtS4Qx6SSpc774e\nwL626qObRm1fd/FxXR6JJM0MZ/SSVDiDXpIKZ9BLUuE6WqOPiKOBazJzWUQcDqwHhoEtwLmZuSci\nzgTOBnYBV2bmnRFxEHALsAgYAlZm5uAM1CFJGkPbGX1EXAjcCMyvm64HVmfmUqAHODUiDgHOA5YA\nJwFXR8Q84BzgkbrvzcDq6S9BkjSeTmb03wXeDHyu3l8M3FtvbwROBHYDmzNzJ7AzIrYCRwLHANe2\n9L28k0H19y+gt3dO234DA32dvN2kzOR7T4f9fXwzqcm1Q7Prb3LtMPn62wZ9Zt4WEYe1NPVk5nC9\nPQQsBA4GnmnpM1r7SFtb27btaNtnYKCPwcGhTt5uUmbyvadqpmvfnzW5dmh2/U2uHTqrf6wTwWQu\nxu5p2e4Dnga219vjtY+0SZK6aDJB/3BELKu3lwP3AQ8CSyNifkQsBI6gulC7GVixV19JUhdNJugv\nAK6IiAeAucCGzPwhsIYqyDcBl2Xms8ANwCsj4n7gLOCK6Rm2JKlTHX29MjOfAF5Tbz8OHDtKn7XA\n2r3adgCnT3mUkqRJ84YpSSqcQS9JhTPoJalwBr0kFc6gl6TCGfSSVDiDXpIKZ9BLUuEMekkqXOP/\ncvCx+JeGSyqFM3pJKpxBL0mFM+glqXAGvSQVzqCXpMIZ9JJUOINekgpn0EtS4Qx6SSqcd8ZOkHfM\nSpptnNFLUuEMekkqnEs308QlHUn7K2f0klQ4g16SCmfQS1LhXKOfYWOt3YPr95K6wxm9JBXOoJek\nwhn0klQ41+j3Q34nX9J0Muj3ofEu1ErSdHHpRpIKZ9BLUuFmfOkmIg4APgkcBewE3peZW2f6c0s0\n0aUe1/QlQXfW6E8D5mfmH0TEa4CPAad24XMbb7pODJ5gpNmtG0F/DHAXQGb+Z0S8ugufqUmYrovD\ns+ki80RPShP9RtR09R/vNdNlpicGTgD2nZ7h4eEZ/YCIuBG4LTM31vvfB16Wmbtm9IMlSUB3LsZu\nB/paP9OQl6Tu6UbQbwZWANRr9I904TMlSbVurNH/M3BCRPwH0AOc0YXPlCTVZnyNXpK0b3nDlCQV\nzqCXpMIZ9JJUuFn19MqmPk4hIo4GrsnMZRFxOLAeGAa2AOdm5p59Ob6ZEhEHAuuAw4B5wJXAozSn\n/jnAWiCo6n0/8CwNqR8gIhYB3wBOAHbRkNoj4ptUX00H+B5wFVOofbbN6J97nAJwMdXjFIoWERcC\nNwLz66brgdWZuZTqW0wlP07incBTda0nA5+gWfWfApCZS4DVVP+zN6b++kT/aeCndVMjao+I+UBP\nZi6r/zmDKdY+24L+Fx6nADThcQrfBd7csr8YuLfe3ggc3/URdc8Xgcvr7R6qGV1j6s/MO4Cz6t2X\nAk/ToPqB64BPAU/W+02p/ShgQUTcExGb6vuPplT7bAv6g4FnWvZ3R8SsWn6aqMy8Dfh5S1NPZo58\nJ3YIWNj9UXVHZv44M4ciog/YQDWrbUz9AJm5KyJuAj4OfJ6G1B8R7wEGM/PuluZG1A7soDrJnUS1\nXDflP/fZFvQ+TgFa1+X6qGZ5xYqIQ4GvAJ/LzC/QsPoBMnMl8FtU6/UHtRwquf5VVDdafhV4FXAz\nsKjleMm1Pw7ckpnDmfk48BTwopbjE659tgW9j1OAhyNiWb29HLhvH45lRkXEi4B7gIsyc13d3KT6\n3xURl9S7O6hOcg81of7MfF1mHpuZy4BvAe8GNjahdqqT3McAIuIlVCsZ90yl9tm27OHjFOACYG1E\nzAUeo1rSKNWlQD9weUSMrNWfD6xpSP23A5+NiK8BBwIfpKq5KX/+e2vKf/ufAdZHxP1U37JZBfyI\nKdTuIxAkqXCzbelGkjRBBr0kFc6gl6TCGfSSVDiDXpIKZ9BLUuEMekkq3P8DW3Ls8EYSqhwAAAAA\nSUVORK5CYII=\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x1ff7cdb50f0>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "plt.figure(figsize=(25, 20))\n",
    "result_genes.regs[result_genes.regs.count_gene_mutation <=50].hist(\"count_gene_mutation\", bins=50)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "3.4381662107348365"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "result_genes.regs['count_gene_mutation'].mean()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>strand</th>\n",
       "      <th>name</th>\n",
       "      <th>score</th>\n",
       "      <th>count_gene_mutation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>S_00000.gdm</th>\n",
       "      <td>chr4</td>\n",
       "      <td>144498560</td>\n",
       "      <td>144621828</td>\n",
       "      <td>-</td>\n",
       "      <td>NM_001168235</td>\n",
       "      <td>0.0</td>\n",
       "      <td>283.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S_00000.gdm</th>\n",
       "      <td>chr15</td>\n",
       "      <td>78463186</td>\n",
       "      <td>78527049</td>\n",
       "      <td>-</td>\n",
       "      <td>NM_001199377</td>\n",
       "      <td>0.0</td>\n",
       "      <td>141.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S_00000.gdm</th>\n",
       "      <td>chr15</td>\n",
       "      <td>78463186</td>\n",
       "      <td>78527049</td>\n",
       "      <td>-</td>\n",
       "      <td>NM_015162</td>\n",
       "      <td>0.0</td>\n",
       "      <td>141.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S_00000.gdm</th>\n",
       "      <td>chr2</td>\n",
       "      <td>27505296</td>\n",
       "      <td>27530307</td>\n",
       "      <td>+</td>\n",
       "      <td>NM_032546</td>\n",
       "      <td>0.0</td>\n",
       "      <td>114.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S_00000.gdm</th>\n",
       "      <td>chr2</td>\n",
       "      <td>27505296</td>\n",
       "      <td>27530307</td>\n",
       "      <td>+</td>\n",
       "      <td>NM_187841</td>\n",
       "      <td>0.0</td>\n",
       "      <td>114.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S_00000.gdm</th>\n",
       "      <td>chr11</td>\n",
       "      <td>116714117</td>\n",
       "      <td>116969131</td>\n",
       "      <td>-</td>\n",
       "      <td>NM_025164</td>\n",
       "      <td>0.0</td>\n",
       "      <td>99.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S_00000.gdm</th>\n",
       "      <td>chr11</td>\n",
       "      <td>116714117</td>\n",
       "      <td>116969131</td>\n",
       "      <td>-</td>\n",
       "      <td>NM_001281749</td>\n",
       "      <td>0.0</td>\n",
       "      <td>99.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S_00000.gdm</th>\n",
       "      <td>chr11</td>\n",
       "      <td>116714117</td>\n",
       "      <td>116969131</td>\n",
       "      <td>-</td>\n",
       "      <td>NM_001281748</td>\n",
       "      <td>0.0</td>\n",
       "      <td>99.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S_00000.gdm</th>\n",
       "      <td>chr4</td>\n",
       "      <td>144257982</td>\n",
       "      <td>144395718</td>\n",
       "      <td>+</td>\n",
       "      <td>NM_002039</td>\n",
       "      <td>0.0</td>\n",
       "      <td>87.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S_00000.gdm</th>\n",
       "      <td>chr4</td>\n",
       "      <td>144257982</td>\n",
       "      <td>144395718</td>\n",
       "      <td>+</td>\n",
       "      <td>NM_207123</td>\n",
       "      <td>0.0</td>\n",
       "      <td>87.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S_00000.gdm</th>\n",
       "      <td>chr15</td>\n",
       "      <td>78730517</td>\n",
       "      <td>78793798</td>\n",
       "      <td>+</td>\n",
       "      <td>NM_004136</td>\n",
       "      <td>0.0</td>\n",
       "      <td>86.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S_00000.gdm</th>\n",
       "      <td>chr16</td>\n",
       "      <td>53737874</td>\n",
       "      <td>54148379</td>\n",
       "      <td>+</td>\n",
       "      <td>NM_001080432</td>\n",
       "      <td>0.0</td>\n",
       "      <td>84.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S_00000.gdm</th>\n",
       "      <td>chr1</td>\n",
       "      <td>218458628</td>\n",
       "      <td>218511325</td>\n",
       "      <td>+</td>\n",
       "      <td>NM_016052</td>\n",
       "      <td>0.0</td>\n",
       "      <td>71.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S_00000.gdm</th>\n",
       "      <td>chr4</td>\n",
       "      <td>88928798</td>\n",
       "      <td>88998931</td>\n",
       "      <td>+</td>\n",
       "      <td>NM_000297</td>\n",
       "      <td>0.0</td>\n",
       "      <td>60.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S_00000.gdm</th>\n",
       "      <td>chr8</td>\n",
       "      <td>2792874</td>\n",
       "      <td>4852328</td>\n",
       "      <td>-</td>\n",
       "      <td>NM_033225</td>\n",
       "      <td>0.0</td>\n",
       "      <td>59.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S_00000.gdm</th>\n",
       "      <td>chr9</td>\n",
       "      <td>21994789</td>\n",
       "      <td>22121093</td>\n",
       "      <td>+</td>\n",
       "      <td>NR_047534</td>\n",
       "      <td>0.0</td>\n",
       "      <td>59.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S_00000.gdm</th>\n",
       "      <td>chr9</td>\n",
       "      <td>21994789</td>\n",
       "      <td>22121093</td>\n",
       "      <td>+</td>\n",
       "      <td>NR_047535</td>\n",
       "      <td>0.0</td>\n",
       "      <td>59.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S_00000.gdm</th>\n",
       "      <td>chr9</td>\n",
       "      <td>21994789</td>\n",
       "      <td>22121093</td>\n",
       "      <td>+</td>\n",
       "      <td>NR_047532</td>\n",
       "      <td>0.0</td>\n",
       "      <td>59.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S_00000.gdm</th>\n",
       "      <td>chr9</td>\n",
       "      <td>21994789</td>\n",
       "      <td>22121093</td>\n",
       "      <td>+</td>\n",
       "      <td>NR_003529</td>\n",
       "      <td>0.0</td>\n",
       "      <td>59.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S_00000.gdm</th>\n",
       "      <td>chr9</td>\n",
       "      <td>21994789</td>\n",
       "      <td>22121093</td>\n",
       "      <td>+</td>\n",
       "      <td>NR_120536</td>\n",
       "      <td>0.0</td>\n",
       "      <td>59.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S_00000.gdm</th>\n",
       "      <td>chr9</td>\n",
       "      <td>21994789</td>\n",
       "      <td>22121093</td>\n",
       "      <td>+</td>\n",
       "      <td>NR_047536</td>\n",
       "      <td>0.0</td>\n",
       "      <td>59.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S_00000.gdm</th>\n",
       "      <td>chr9</td>\n",
       "      <td>21994789</td>\n",
       "      <td>22121093</td>\n",
       "      <td>+</td>\n",
       "      <td>NR_047538</td>\n",
       "      <td>0.0</td>\n",
       "      <td>59.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S_00000.gdm</th>\n",
       "      <td>chr9</td>\n",
       "      <td>21994789</td>\n",
       "      <td>22121093</td>\n",
       "      <td>+</td>\n",
       "      <td>NR_047543</td>\n",
       "      <td>0.0</td>\n",
       "      <td>59.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S_00000.gdm</th>\n",
       "      <td>chr9</td>\n",
       "      <td>21994789</td>\n",
       "      <td>22121093</td>\n",
       "      <td>+</td>\n",
       "      <td>NR_047537</td>\n",
       "      <td>0.0</td>\n",
       "      <td>59.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S_00000.gdm</th>\n",
       "      <td>chr15</td>\n",
       "      <td>78556486</td>\n",
       "      <td>78574538</td>\n",
       "      <td>+</td>\n",
       "      <td>NM_018602</td>\n",
       "      <td>0.0</td>\n",
       "      <td>52.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S_00000.gdm</th>\n",
       "      <td>chr9</td>\n",
       "      <td>133028157</td>\n",
       "      <td>133309510</td>\n",
       "      <td>+</td>\n",
       "      <td>NM_001291815</td>\n",
       "      <td>0.0</td>\n",
       "      <td>51.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               chr      start       stop strand          name  score  \\\n",
       "S_00000.gdm   chr4  144498560  144621828      -  NM_001168235    0.0   \n",
       "S_00000.gdm  chr15   78463186   78527049      -  NM_001199377    0.0   \n",
       "S_00000.gdm  chr15   78463186   78527049      -     NM_015162    0.0   \n",
       "S_00000.gdm   chr2   27505296   27530307      +     NM_032546    0.0   \n",
       "S_00000.gdm   chr2   27505296   27530307      +     NM_187841    0.0   \n",
       "S_00000.gdm  chr11  116714117  116969131      -     NM_025164    0.0   \n",
       "S_00000.gdm  chr11  116714117  116969131      -  NM_001281749    0.0   \n",
       "S_00000.gdm  chr11  116714117  116969131      -  NM_001281748    0.0   \n",
       "S_00000.gdm   chr4  144257982  144395718      +     NM_002039    0.0   \n",
       "S_00000.gdm   chr4  144257982  144395718      +     NM_207123    0.0   \n",
       "S_00000.gdm  chr15   78730517   78793798      +     NM_004136    0.0   \n",
       "S_00000.gdm  chr16   53737874   54148379      +  NM_001080432    0.0   \n",
       "S_00000.gdm   chr1  218458628  218511325      +     NM_016052    0.0   \n",
       "S_00000.gdm   chr4   88928798   88998931      +     NM_000297    0.0   \n",
       "S_00000.gdm   chr8    2792874    4852328      -     NM_033225    0.0   \n",
       "S_00000.gdm   chr9   21994789   22121093      +     NR_047534    0.0   \n",
       "S_00000.gdm   chr9   21994789   22121093      +     NR_047535    0.0   \n",
       "S_00000.gdm   chr9   21994789   22121093      +     NR_047532    0.0   \n",
       "S_00000.gdm   chr9   21994789   22121093      +     NR_003529    0.0   \n",
       "S_00000.gdm   chr9   21994789   22121093      +     NR_120536    0.0   \n",
       "S_00000.gdm   chr9   21994789   22121093      +     NR_047536    0.0   \n",
       "S_00000.gdm   chr9   21994789   22121093      +     NR_047538    0.0   \n",
       "S_00000.gdm   chr9   21994789   22121093      +     NR_047543    0.0   \n",
       "S_00000.gdm   chr9   21994789   22121093      +     NR_047537    0.0   \n",
       "S_00000.gdm  chr15   78556486   78574538      +     NM_018602    0.0   \n",
       "S_00000.gdm   chr9  133028157  133309510      +  NM_001291815    0.0   \n",
       "\n",
       "             count_gene_mutation  \n",
       "S_00000.gdm                283.0  \n",
       "S_00000.gdm                141.0  \n",
       "S_00000.gdm                141.0  \n",
       "S_00000.gdm                114.0  \n",
       "S_00000.gdm                114.0  \n",
       "S_00000.gdm                 99.0  \n",
       "S_00000.gdm                 99.0  \n",
       "S_00000.gdm                 99.0  \n",
       "S_00000.gdm                 87.0  \n",
       "S_00000.gdm                 87.0  \n",
       "S_00000.gdm                 86.0  \n",
       "S_00000.gdm                 84.0  \n",
       "S_00000.gdm                 71.0  \n",
       "S_00000.gdm                 60.0  \n",
       "S_00000.gdm                 59.0  \n",
       "S_00000.gdm                 59.0  \n",
       "S_00000.gdm                 59.0  \n",
       "S_00000.gdm                 59.0  \n",
       "S_00000.gdm                 59.0  \n",
       "S_00000.gdm                 59.0  \n",
       "S_00000.gdm                 59.0  \n",
       "S_00000.gdm                 59.0  \n",
       "S_00000.gdm                 59.0  \n",
       "S_00000.gdm                 59.0  \n",
       "S_00000.gdm                 52.0  \n",
       "S_00000.gdm                 51.0  "
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "result_genes.regs[result_genes.regs.count_gene_mutation > 50].sort_values(\"count_gene_mutation\", ascending=False)"
   ]
  }
 ],
 "metadata": {
  "anaconda-cloud": {},
  "kernelspec": {
   "display_name": "Python (ml_experiments)",
   "language": "python",
   "name": "ml_experiments"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
